hoomd/mpcd/RejectionVirtualParticleFillerGPU.h

```diff
@@ -0,0 +1,189 @@
+    };
+
+
+template<class Geometry>
+void RejectionVirtualParticleFillerGPU<Geometry>::fill(unsigned int timestep)
+    {
+    // Number of particles that we need to draw (constant)
+    const BoxDim& box = this->m_pdata->getBox();
+    const Scalar3 lo = box.getLo();
+    const Scalar3 hi = box.getHi();
+    const unsigned int N_virt_max = round(this->m_density*box.getVolume());
+
+    // Step 1: Create temporary arrays
+    if (N_virt_max > this->m_tmp_pos.getNumElements())
+        {
+        GPUArray<Scalar4> tmp_pos(N_virt_max, this->m_exec_conf);
+        this->m_tmp_pos.swap(tmp_pos);
+        GPUArray<Scalar4> tmp_vel(N_virt_max, this->m_exec_conf);
+        this->m_tmp_vel.swap(tmp_vel);
+        GPUArray<bool> track_bounded_particles(N_virt_max, this->m_exec_conf);
+        m_track_bounded_particles.swap(track_bounded_particles);
+        GPUArray<unsigned int> compact_idxs(N_virt_max, this->m_exec_conf);
+        m_compact_idxs.swap(compact_idxs);
+        GPUFlags<unsigned int> num_selected(this->m_exec_conf);
+        m_num_selected.swap(num_selected);
+        }
+    ArrayHandle<Scalar4> d_tmp_pos(this->m_tmp_pos, access_location::device, access_mode::overwrite);
+    ArrayHandle<Scalar4> d_tmp_vel(this->m_tmp_vel, access_location::device, access_mode::overwrite);
+    ArrayHandle<bool> d_track_bounded_particles(m_track_bounded_particles, access_location::device, access_mode::overwrite);
+    ArrayHandle<unsigned int> d_compact_idxs(m_compact_idxs, access_location::device, access_mode::overwrite);
+
+    // Step 2: Draw particle positions and velocities in parallel on GPU
+    unsigned int first_tag = this->computeFirstTag(N_virt_max);
+
+    const Scalar vel_factor = fast::sqrt(this->m_T->getValue(timestep) / this->m_mpcd_pdata->getMass());
+
+    mpcd::gpu::draw_virtual_particles_args_t args(d_tmp_pos.data,
+                                                  d_tmp_vel.data,
+                                                  d_track_bounded_particles.data,
+                                                  lo,
+                                                  hi,
+                                                  first_tag,
+                                                  vel_factor,
+                                                  this->m_filler_id,
+                                                  this->m_type,
+                                                  N_virt_max,
+                                                  timestep,
+                                                  this->m_seed,
+                                                  m_tuner1->getParam());
+
+    // Pick (N=box_vol*solvPartDensity) particles and velocities randomly in the box
+    m_tuner1->begin();
+    mpcd::gpu::draw_virtual_particles<Geometry>(args, *(this->m_geom));
+    if (this->m_exec_conf->isCUDAErrorCheckingEnabled()) CHECK_CUDA_ERROR();
+    m_tuner1->end();
+
+
+    // compact particle indices
```

Comment:
You should indent and encapsulate all of this (line 130-149) in curly brackets, that way `d_tmp_alloc` will go out of scope after CUB is done using it.